messenger/message-recorder.rst

```diff
@@ -0,0 +1,120 @@
+dispatches a command named ``SignUpUser`` to the command bus. The command is handled by the
+``SignUpUserHandler`` which creates a ``User`` object, stores that object to a database and
+dispatches a ``UserSignedUp`` event to the event bus.
+
+There are many subscribers to the ``UserSignedUp`` event, one subscriber may send
+a welcome email to the new user. We are using the ``DoctrineTransactionMiddleware``
+to wrap all database queries in one database transaction.
+
+**Problem 1:** If an exception is thrown when sending the welcome email, then the user
+will not be created because the ``DoctrineTransactionMiddleware`` will rollback the
+Doctrine transaction, in which the user has been created.
+
+**Problem 2:** If an exception is thrown when saving the user to the database, the welcome
+email is still sent.
+
+``HandleMessageInNewTransaction`` Middleware
+--------------------------------------------
+
+For many applications, the desired behavior is to have any messages dispatched by the handler
+to `only` be handled after the handler finishes. This can be by using the
+``HandleMessageInNewTransaction`` middleware and adding a ``Transaction`` stamp to
+`the message Envelope </components/messenger#adding-metadata-to-messages-envelopes>`_.
+This middleware enables us to add messages to a separate transaction that will only be
+dispatched *after* the current message handler finishes.
+
+Referencing the above example, this means that the ``UserSignedUp`` event would not be handled
+until *after* the ``SignUpUserHandler`` had completed and the new ``User`` was persisted to the
+database. If the ``SignUpUserHandler`` encounters an exception, the ``UserSignedUp`` event will
+never be handled and if an exception is thrown while sending the welcome email, the Doctrine
+transaction will not be rolled back.
+
+To enable this, you need to add the ``handle_message_in_new_transaction``
+middleware. Make sure it is registered before ``DoctrineTransactionMiddleware``
+in the middleware chain.
+
+**Note:** The ``handle_message_in_new_transaction`` middleware must be loaded for *all* of the
+buses. For the example, the middleware must be loaded for both the command and event buses.
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/packages/messenger.yaml
+        framework:
+            messenger:
+                default_bus: messenger.bus.command
+
+                buses:
+                    messenger.bus.command:
+                        middleware:
+                            - validation
+                            - handle_message_in_new_transaction
+                            - doctrine_transaction
+                    messenger.bus.event:
+                        default_middleware: allow_no_handlers
+                        middleware:
+                            - validation
+                            - handle_message_in_new_transaction
+                            - doctrine_transaction
```

Comment:
doctrine transaction middleware in the event bus isn't a necessity. I would not add it here as it's only used in the command bus to demonstrate the events are only dispatched after the doctrine transaction is committed.
The event may not even use doctrine, not have a handler, or could just be a message sent asynchronously to a queue. So what produces the event here is another matter and the `doctrine_transaction` here might only bring confusion to the reader.

I'm not sure either about validation, as you're the one creating the event, I guess this would be a logic error from the developper if the crafted event is invalid somehow.

Reply:
That makes sense - even if they may be useful in some people's applications, there is no reason to have them in example code.

- The middleware order for the command bus shows that the `handle_message_in_new_transaction` must be before `doctine_transaction`
- The `handle_message_in_new_transaction` being declared for both buses shows that it must be declared of all involved buses.

Since both notes/warning are shown in the example, no reason to have extra middlewares.

```suggestion
                        middleware:
                            - handle_message_in_new_transaction
```

should be sufficient for the example

Reply:
add it commented out with a description above:

```
> # uncomment if you want each message to be wrapped in a transaction
> # when used with handle_message_in_new_transaction, each message will be wrapped
> # in their own, separate transactions
```

Reply:
I agree, we could remove it.